Add nix flake

theoparis opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

Based on #154, this PR adds a nix flake.
The changes I've made compared to #154 are:
- Added a Nix derivation
- Added support for other systems (e.g. riscv64) via nix-systems
- Used nixfmt (nixfmt-rfc-style package) for formatting

Darwin support is still WIP which is why I made this a draft. It seems like AppKit/AppKit.h is not being found despite it being present in buildInputs. 
Additionally, I had to temporarily modify platform_methods.py to allow finding moltenVK in `${darwin.moltenvk}/lib/libMoltenVK.dylib`. This could probably be improved with a proper lookup of the entire sdk directory instead of looking in lib or other hard coded paths 🤔

Spartan322 commented:
Is this still being worked on, and has this been proposed to Godot?

Spartan322 commented:
Also does this close #154?

MichaelFisher1997 commented:
@Spartan322 I have my eye on this.
#154 is mine, this is an improvement on that (so probs close #154)
its gone a bit far past my knowledge of nix shells, but a few people in the community are asking for it.
ive spoken to a few people, we will be getting back to this, a few use it already to do local builds

MichaelFisher1997 commented:
> Is this still being worked on, and has this been proposed to Godot?

yes, no

MichaelFisher1997 commented:
can we maybe move this to https://github.com/Redot-Engine/redot-proposals/?

theoparis commented:
An update on this PR: 
- I moved to nixpkgs master instead of flakehub, however if preferred I can switch to nixpkgs-unstable. Personally I think the master branch is still stable enough while receiving the latest updates from nixpkgs. Due to using flakes there is a lock file that pins a specific commit of nixpkgs.
- I added Dotnet 9 to the package.nix file, along with a enableMono option that is currently untested. I think there are some post compilation steps that I haven't added yet for that to work. I'm also not sure if Redot supports Dotnet 9 yet.
- I added runtimeDependencies with libGL so that OpenGL/EGL and the rest of the shared libraries can be found.

I recently installed NixOS via dual booting on my new macbook pro so I haven't really worked on Darwin support lately - I assume it's still broken.

Arctis-Fireblight commented:
I think at this stage, we are going to close out this PR. If we end up moving forward with Nix packaging, it will likely be with #954
